packager/packager.cc

```diff
@@ -1676,6 +1677,75 @@ class PackageInfoFormatter final {
 
 } // namespace
 
+absl::Span<core::FileRef> Packager::jezPrototype(const core::GlobalState &gs, absl::Span<core::FileRef> files,
+                                                 const vector<string> &rootPackages) {
+    if (rootPackages.empty()) {
+        // TODO(jez) Build this into a proper mode, where this doesn't just print things
+        return files;
+    }
+
+    // TODO(jez) Don't have package IDs, so we can't easily make this vector<bool>
+    auto requiredPackages = UnorderedSet<core::NameRef>{};
+
+    queue<core::NameRef> bfsFrontier;
+    // TODO(jez) the name root package is bad, because "root package" has another meaning (the root
+    // of a workspace, not the root of a BFS search)
+    for (const auto &rootPackageName : rootPackages) {
+        auto &rootPackage = gs.packageDB().getPackageInfo(gs, rootPackageName);
+        if (!rootPackage.exists()) {
+            Exception::raise("Unknown package: {}", rootPackageName);
+        }
+
+        bfsFrontier.emplace(rootPackage.mangledName());
+    }
+
+    // TODO(jez) Can parallelize BFS if you need
+    while (!bfsFrontier.empty()) {
+        auto curPackageName = bfsFrontier.front();
+        bfsFrontier.pop();
+
+        if (requiredPackages.find(curPackageName) != requiredPackages.end()) {
+            continue;
+        }
+        requiredPackages.emplace(curPackageName);
+
+        auto &curPackageAbstract = gs.packageDB().getPackageInfo(curPackageName);
+        auto &curPackage = PackageInfoImpl::from(curPackageAbstract);
+
+        for (auto &importedPackageName : curPackage.importedPackageNames) {
+            if (importedPackageName.type != ImportType::Normal) {
+                // TODO(jez) Figure out what to do about test code
+                continue;
+            }
+            if (!importedPackageName.name.mangledName.exists()) {
+                // TODO(jez) Does this happen? Error recovery?
+                continue;
+            }
```

Comment:
Probably should `ENFORCE` this case, since it shouldn't happen

Reply:
will do—this PR is very wip